perception/sandbox/polygrasp/src/polygrasp/grasp_rpc.py

```diff
@@ -0,0 +1,178 @@
+
+
+log = logging.getLogger(__name__)
+topic_key = "grasp_server"
+grasp_topic_key = f"{topic_key}/grasp"
+collision_topic_key = f"{topic_key}/collision"
+
+
+def save_img(img, name):
+    f = plt.figure()
+    plt.imshow(img)
+    f.savefig(f"{name}.png")
+    plt.close(f)
+
+
+class GraspServer:
+    def _get_grasps(self, pcd: o3d.geometry.PointCloud) -> np.ndarray:
+        raise NotImplementedError
+
+    def _get_collisions(
+        self, grasps: graspnetAPI.GraspGroup, scene_pcd: o3d.geometry.PointCloud
+    ) -> graspnetAPI.GraspGroup:
+        raise NotImplementedError
+
+    def start(self):
+        log.info(f"Starting grasp server...")
+
+        def grasp_onrequest(req):
+            log.info(f"Got request; computing grasp group...")
+
+            payload = req.pkt.payload
+            pcd = serdes.capnp_to_pcd(payload)
+            grasp_group = self._get_grasps(pcd)
+
+            log.info(f"Done. Replying with serialized grasp group...")
+            req.reply(serdes.grasp_group_to_capnp(grasp_group).to_bytes())
+
+        grasp_server = a0.RpcServer(grasp_topic_key, grasp_onrequest, None)
+
+        def collision_onrequest(req):
+            log.info(f"Got request; computing collisions...")
+
+            payload = req.pkt.payload
+            msg = polygrasp_msgs.CollisionRequest.from_bytes(payload)
+            grasp_group = serdes.bytes_to_grasp_group(msg.grasps)
+            scene_pcd = serdes.capnp_to_pcd(msg.pcd)
+
+            filtered_grasp_group = self._get_collisions(grasp_group, scene_pcd)
+            log.info(f"Done. Replying with serialized filtered grasps...")
+            req.reply(serdes.grasp_group_to_bytes(filtered_grasp_group))
+
+        collision_server = a0.RpcServer(collision_topic_key, collision_onrequest, None)
+
+        while True:
+            pass
+
+
+class GraspClient:
+    def __init__(self, view_json_path):
+        self.grasp_client = a0.RpcClient(grasp_topic_key)
+        self.collision_client = a0.RpcClient(collision_topic_key)
+        self.view_json_path = view_json_path
+
+    def downsample_pcd(self, pcd: o3d.geometry.PointCloud, max_num_bits=8 * 1024 * 1024):
+        # a0 default max msg size 16MB; make sure every msg < 1/2 of max
+        i = 1
+        while True:
+            downsampled_pcd = pcd.uniform_down_sample(i)
+            bits = serdes.pcd_to_capnp(downsampled_pcd).to_bytes()
+            if len(bits) > max_num_bits:
+                log.warning(f"Downsampling pointcloud...")
+                i += 1
+            else:
+                break
+        if i > 1:
+            log.warning(f"Downsampled to every {i}th point.")
+
+        return bits
+
+    def get_grasps(self, pcd: o3d.geometry.PointCloud) -> graspnetAPI.GraspGroup:
+        state = []
+
+        def onreply(pkt):
+            state.append(pkt.payload)
+
+        bits = self.downsample_pcd(pcd)
+        self.grasp_client.send(bits, onreply)
+
+        while not state:
+            pass
```

Comment:
See above comment about burning cpu.

Reply:
I believe there is a `send_blocking` method of the RpcClient, which will let you avoid this.